### PR TITLE
client: do not raise an exception for submit_sm_resp

### PR DESCRIPTION
**Pull request by novns**
Allow the handler to use error status from the PDU

### CODE REVIEW

**Comment by podshumok** [comment]
Hello!
Could you please describe what are you trying to do and why?

**Comment by novns** [comment]
I'm trying to get full information available in submit_sm_resp PDU, including pdu.sequence and pdu.status, in my sent_handler function:

    def sent_handler(self, pdu):
        if pdu.status:
            L.error('Cannot send part %i. Error %i' % (pdu.sequence, pdu.status))
            # useful error handling here
            return

        L.debug('Sent part %i' % pdu.sequence)
        # useful success handling here

Currently sent_handler() is never called back if SMSC returns any error, because unneeded exception is raised.

**Comment by podshumok** [comment]
may be it would be better to put that information into the exception?

**Comment by novns** [comment]
It will be useful too, but inconsistent a little.
The submit_sm_resp with an error status is not any kind of broken PDU, it is the same normal responce.

I don't see why we have to process the similar responce in two different places with the less readable code.

BTW I didn't look at other cases of p.command, but is this exception really necessary at all?

**Comment by podshumok** [comment]
`PDUError` currently only raised from client if `pdu.status` is not success.

What you can do:
a) Add `raise_for_status` option to the client (defaults to `True`),  and `raise_for_status`  method to `PDU`. If the flag is set, than client should call `pdu.rasie_for_status` instead of checking the status itself.
When flag is off handlers are responsible for doing it right for error statuses.

b) Add additional information to the `PDUError` exception. `status` is already there, yo may want to add sequence number or pdu object itself.

**Comment by novns** [comment]
It looks like the whole library needs a total refactoring and rewriting.

Just made an ad-hoc fix for another annoying bug: https://github.com/podshumok/python-smpplib/pull/35/commits/31a02640ce848b202883f16827763e98206676e1

File: smpplib/command.py [hunk 1]
```diff
@@ -148,7 +148,7 @@ def generate_params(self):
                     value = self._generate_int(field)
                     body += value
                 elif param.type is str:
-                    value = self._generate_string(field)
+                    value = self._generate_string(field).encode()
```
Comment (podshumok) [inline smpplib/command.py]: Hello!
Thank you for finding this one. We've moved to python3 recently and while some people use the library in production, there still could be some issues.
Please open separate PR for that patch

File: smpplib/client.py [hunk 2]
```diff
@@ -283,7 +283,7 @@ def read_once(self, ignore_error_codes=None):
                 self.send_pdu(p)
                 return
 
-            if p.is_error():
+            if p.command != 'submit_sm_resp' and p.is_error():
```
Comment (eigenein) [inline smpplib/client.py]: This is too specific to the particular type of PDU and breaks the general flow. I rather like @podshumok's suggestion to add `raise_for_status` or something like that.

File: smpplib/command.py (hunk 1 above)
Comment (eigenein) [reply smpplib/command.py]: `_generate_string` returns a byte-string already: `return six.b(value)`

**Pull request closed without merging**